gnu_dataset/mifluz-0.26.0/man/WordDict.3

'''
''' Part of the ht://Dig package   <http://www.htdig.org/>
''' Copyright (c) 1999, 2000, 2001 The ht://Dig Group
''' For copyright details, see the file COPYING in your distribution
''' or the GNU General Public License version 2 or later
''' <http://www.gnu.org/copyleft/gpl.html>
''' 
''' 
.TH WordDict 3 local
.SH NAME
WordDict \-

manage and use an inverted index dictionary.


.SH SYNOPSIS
.nf
.ft CW

#include <mifluz.h>

WordList* words = ...;
WordDict* dict = words->Dict();
.ft R
.fi

.SH DESCRIPTION

WordDict maps strings to unique identifiers and frequency in the 
inverted index. Whenever a new word is found, the WordDict class 
can be asked to assign it a serial number. When doing so, an entry
is created in the dictionary with a frequency of zero. The application
may then increment or decrement the frequency to reflect the inverted
index content.

The serial numbers range from 1 to 2^32 inclusive.

A WordDict object is automatically created by the WordList object and
should not be created directly by the application.


.SH METHODS
.TP
.B   WordDict() 
Private constructor.
.TP
.B   int Initialize(WordList* words)
Bind the object a WordList inverted index. Return OK on success,
NOTOK otherwise.
.TP
.B   int Open()
Open the underlying Berkeley DB sub-database. The enclosing 
file is given by the
.I words
data member. Return OK on success,
NOTOK otherwise.
.TP
.B   int Remove()
Destroy the underlying Berkeley DB sub-database. Return OK on success,
NOTOK otherwise.
.TP
.B   int Close()
Close the underlying Berkeley DB sub-database. Return OK on success,
NOTOK otherwise.
.TP
.B   int Serial(const String& word, unsigned int& serial)
If the
.B word
argument exists in the dictionnary, return its
serial number in the
.B serial
argument. If it does not already
exists, assign it a serial number, create an entry with a frequency
of zero and return the new serial in the
.B serial
argument.
Return OK on success, NOTOK otherwise.
.TP
.B   int SerialExists(const String& word, unsigned int& serial)
If the
.B word
argument exists in the dictionnary, return its
serial number in the
.B serial
argument. If it does not exists
set the
.B serial
argument to WORD_DICT_SERIAL_INVALID.
Return OK on success, NOTOK otherwise.
.TP
.B   int SerialRef(const String& word, unsigned int& serial)
Short hand for Serial() followed by Ref().
Return OK on success, NOTOK otherwise.
.TP
.B   int Noccurrence(const String& word, unsigned int& noccurrence) const
Return the frequency of the
.B word
argument
in the
.B noccurrence
argument. 
Return OK on success, NOTOK otherwise.
.TP
.B   int Normalize(String& word) const
Short hand for words->GetContext()->GetType()->Normalize(word).
Return OK on success, NOTOK otherwise.
.TP
.B   int Ref(const String& word) 
Short hand for Incr(word, 1)
.TP
.B   int Incr(const String& word, unsigned int incr)
Add
.B incr
to the frequency of the
.B word
. 
Return OK on success, NOTOK otherwise.
.TP
.B   int Unref(const String& word) 
Short hand for Decr(word, 1)
.TP
.B   int Decr(const String& word, unsigned int decr)
Subtract
.B decr
to the frequency of the
.B word
. If
the frequency becomes lower or equal to zero, remove the entry
from the dictionnary and lose the association between the word and its
serial number.
Return OK on success, NOTOK otherwise.
.TP
.B   int Put(const String& word, unsigned int noccurrence)
Set the frequency of
.B word
with the value of the
.B noccurrence
argument.
.TP
.B   int Exists(const String& word) const
Return true if
.B word
exists in the dictionnary, false otherwise.
.TP
.B   List* Words() const
Return a pointer to the associated WordList object.
.TP
.B   WordDictCursor* Cursor() const
Return a cursor to sequentially walk the dictionnary using the
.B Next
method.
.TP
.B   int Next(WordDictCursor* cursor, String& word, WordDictRecord& record)
Return the next entry in the dictionnary. The
.B cursor
argument
must have been created using the
.I Cursor
method. The word is
returned in the
.B word
argument and the record is returned in
the
.B record
argument. 
On success the function returns 0, at the end of the dictionnary it
returns DB_NOTFOUND. The
.B cursor
argument is deallocated when
the function hits the end of the dictionnary or an error occurs.
.TP
.B   WordDictCursor* CursorPrefix(const String& prefix) const
Return a cursor to sequentially walk the entries of the dictionnary
that start with the
.B prefix
argument, using the
.B NextPrefix
method.
.TP
.B   int NextPrefix(WordDictCursor* cursor, String& word, WordDictRecord& record)
Return the next prefix from the dictionnary. The
.B cursor
argument
must have been created using the
.I CursorPrefix
method. The word is
returned in the
.B word
argument and the record is returned in
the
.B record
argument. The
.B word
is guaranteed to start with
the prefix specified to the
.B CursorPrefix
method.
On success the function returns 0, at the end of the dictionnary it
returns DB_NOTFOUND. The
.B cursor
argument is deallocated when
the function hits the end of the dictionnary or an error occurs.
.TP
.B   int Write(FILE* f)
Dump the complete dictionary in the file descriptor
.B f.
The
format of the dictionary is
.I word serial frequency
, one by
line.
.PP

.SH AUTHORS
Loic Dachary loic@gnu.org

The Ht://Dig group http://dev.htdig.org/


.SH SEE ALSO
htdb_dump(1), htdb_stat(1), htdb_load(1), mifluzdump(1), mifluzload(1), mifluzsearch(1), mifluzdict(1), WordContext(3), WordList(3), WordListOne(3), WordKey(3), WordKeyInfo(3), WordType(3), WordDBInfo(3), WordRecordInfo(3), WordRecord(3), WordReference(3), WordCursor(3), WordCursorOne(3), WordMonitor(3), Configuration(3), mifluz(3)



gnu_dataset/mifluz-0.26.0/mifluz/WordDict.cc
//
// WordDict.cc
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999, 2000, 2001 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU General Public License version 2 or later
// <http://www.gnu.org/copyleft/gpl.html>
//
// $Id: WordDict.cc,v 1.9 2001/06/29 14:14:08 loic Exp $
//
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <fcntl.h>

#include "WordDict.h"
#include "WordListOne.h"

#define WORD_DICT_CURSOR_FIRST	1
#define WORD_DICT_CURSOR_NEXT	2

class WordDictCursor {
public:
  int info;
  String prefix;
  WordDBCursor* cursor;
};

WordDict::~WordDict()
{
  delete db;
}

int WordDict::Initialize(WordList* nwords)
{
  words = nwords;
  db = new WordDB(nwords->GetContext()->GetDBInfo());
  return OK;
}

int WordDict::Open()
{
  const String& filename = words->Filename();
  int flags = words->Flags();

  db->set_pagesize(words->Pagesize());

  return db->Open(filename, "dict", DB_BTREE, flags, 0666, WORD_DB_DICT) == 0 ? OK : NOTOK;
}

int WordDict::Remove()
{
  return db->Remove(words->Filename(), "dict") == 0 ? OK : NOTOK;
}

int WordDict::Close()
{
  return db->Close() == 0 ? OK : NOTOK;
}

int WordDict::Serial(const String& word, unsigned int& serial)
{
  int ret;
  WordDictRecord entry;
  if((ret = entry.Get(db, word)) != 0 && ret != DB_NOTFOUND)
    return NOTOK;
  if(ret == DB_NOTFOUND) {
    words->Meta()->Serial(WORD_META_SERIAL_WORD, entry.id);
    if(entry.Put(db, word) != 0) return NOTOK;
  }
  serial = entry.id;

  return OK;
}

int WordDict::SerialExists(const String& word, unsigned int& serial)
{
  int ret;
  WordDictRecord entry;
  if((ret = entry.Get(db, word)) != 0 && ret != DB_NOTFOUND)
    return NOTOK;

  serial = ret == DB_NOTFOUND ? WORD_DICT_SERIAL_INVALID : entry.id;

  return OK;
}

int WordDict::SerialRef(const String& word, unsigned int& serial)
{
  int ret;
  WordDictRecord entry;
  if((ret = entry.Get(db, word)) != 0 && ret != DB_NOTFOUND)
    return NOTOK;
  if(ret == DB_NOTFOUND)
    words->Meta()->Serial(WORD_META_SERIAL_WORD, entry.id);
  entry.count++;
  if(entry.Put(db, word) != 0) return NOTOK;
  serial = entry.id;

  return OK;
}

int WordDict::Noccurrence(const String& word, unsigned int& noccurrence) const
{
  if(word.empty()) {
    fprintf(stderr, "WordDict::Noccurrence: null word\n");
    return NOTOK;
  }
  WordDictRecord entry;
  noccurrence = 0;
  int ret;
  if((ret = entry.Get(db, word)) != 0) {
    if(ret != DB_NOTFOUND)
      return NOTOK;
  }
  noccurrence = entry.count;

  return OK;
}

int WordDict::Normalize(String& word) const
{
  const WordType& wtype = words->GetContext()->GetType();

  return wtype.Normalize(word);
}

int WordDict::Incr(const String& word, unsigned int incr)
{
  int ret;
  WordDictRecord entry;
  if((ret = entry.Get(db, word)) != 0 && ret != DB_NOTFOUND)
    return NOTOK;
  if(ret == DB_NOTFOUND)
    words->Meta()->Serial(WORD_META_SERIAL_WORD, entry.id);
  entry.count += incr;
  if(entry.Put(db, word) != 0) return NOTOK;
  return OK;
}

int WordDict::Decr(const String& word, unsigned int decr)
{
  WordDictRecord entry;
  int ret;
  if((ret = entry.Get(db, word)) != 0) {
    if(ret == DB_NOTFOUND)
      fprintf(stderr, "WordDict::Unref(%s) Unref on non existing word occurrence\n", (const char*)word);
    return NOTOK;
  }
  entry.count -= decr;
  if(entry.count > 0)
    ret = entry.Put(db, word) == 0 ? OK : NOTOK;
  else
    ret = entry.Del(db, word) == 0 ? OK : NOTOK;

  return ret;
}

int WordDict::Put(const String& word, unsigned int noccurrence)
{
  int ret;
  WordDictRecord entry;
  if((ret = entry.Get(db, word)) != 0 && ret != DB_NOTFOUND)
    return NOTOK;
  if(ret == DB_NOTFOUND)
    words->Meta()->Serial(WORD_META_SERIAL_WORD, entry.id);
  entry.count = noccurrence;
  if(entry.Put(db, word) != 0) return NOTOK;
  return OK;
}

List *WordDict::Words() const
{
  String key;
  String coded;
  WordDBCursor* cursor = db->Cursor();
  List* list = new List;

  while(cursor->Get(key, coded, DB_NEXT) == 0)
    list->Add(new String(key));

  delete cursor;

  return list;
}

int WordDict::Exists(const String& word) const
{
  String tmp_word = word;
  String coded;
  
  return db->Get(0, tmp_word, coded, 0) == 0;
}

WordDictCursor* WordDict::Cursor() const
{
  WordDictCursor* cursor = new WordDictCursor;
  cursor->cursor = db->Cursor();

  return cursor;
}

int WordDict::Next(WordDictCursor* cursor, String& word, WordDictRecord& record)
{
  String coded;
  int ret = cursor->cursor->Get(word, coded, DB_NEXT);
  if(ret != 0) {
    delete cursor->cursor;
    delete cursor;
  } else {
    record.Unpack(coded);
  }
  return ret;
}

WordDictCursor* WordDict::CursorPrefix(const String& prefix) const
{
  WordDictCursor* cursor = new WordDictCursor;
  cursor->cursor = db->Cursor();
  cursor->prefix = prefix;
  cursor->info = WORD_DICT_CURSOR_FIRST;

  return cursor;
}

int WordDict::NextPrefix(WordDictCursor* cursor, String& word, WordDictRecord& record)
{
  String coded;
  int ret;
  if(cursor->info == WORD_DICT_CURSOR_FIRST) {
    word = cursor->prefix;
    ret = cursor->cursor->Get(word, coded, DB_SET_RANGE);
    cursor->info = WORD_DICT_CURSOR_NEXT;
  } else {
    ret = cursor->cursor->Get(word, coded, DB_NEXT);
  }
  //
  // Stop walking when 1) DB_NOTFOUND, 2) the word found is shorter than
  // the required prefix, 3) the word found does not start with the 
  // required prefix.
  //
  if(ret != 0 ||
     cursor->prefix.length() > word.length() ||
     strncmp(cursor->prefix.get(), word.get(), cursor->prefix.length())) {
    delete cursor->cursor;
    delete cursor;
    if(ret == 0) ret = DB_NOTFOUND;
  } else {
    record.Unpack(coded);
  }
  return ret;
}

int WordDict::Write(FILE* f)
{
  WordDBCursor* cursor = db->Cursor();
  String key;
  String coded;
  unsigned int occurrence;
  unsigned int id;

  while(cursor->Get(key, coded, DB_NEXT) == 0) {
    int offset = 0;
    coded.ber_shift(offset, occurrence);
    coded.ber_shift(offset, id);
    fprintf(f, "%s %d %d\n", (char*)key, id, occurrence);
  }

  delete cursor;

  return OK;
}
